src/synapse_mcp/auth.py

```diff
@@ -1,98 +1,169 @@
-        """
-        auth_url = f"https://signin.synapse.org?{urlencode({'client_id': client_id, 'redirect_uri': redirect_uri, 'response_type': 'code', 'scope': f'openid {scope}'})}"
-        return auth_url
-    def get_client(self) -> synapseclient.Synapse:
-        """Get the authenticated Synapse client."""
-        if not self.synapse_client:
-            raise RuntimeError("Synapse client not initialized")
+import base64
+import json
+import logging
+import asyncio
+from typing import Optional, Dict, Any, List
+from fastmcp.server.auth import OAuthProxy
+
+# Set up logging for token debugging
+logger = logging.getLogger("synapse_mcp.auth")
+
+class SynapseJWTVerifier:
+    """Custom JWT verifier for Synapse tokens that returns FastMCP-compatible results"""
+
+    def __init__(self, jwks_uri: str, issuer: str, audience: str, algorithm: str = "RS256", required_scopes: Optional[List[str]] = None):
+        self.jwks_uri = jwks_uri
+        self.issuer = issuer
+        self.audience = audience
+        self.algorithm = algorithm
+        self.required_scopes = required_scopes or []
+        self._jwks_cache = None
+
+    async def _get_jwks(self):
+        """Get JWKS from Synapse endpoint"""
+        if self._jwks_cache is None:
+            import httpx
+            async with httpx.AsyncClient() as client:
+                response = await client.get(self.jwks_uri)
+                self._jwks_cache = response.json()
+        return self._jwks_cache
+
+    async def verify_token(self, token: str) -> Optional[Dict[str, Any]]:
+        """Verify Synapse JWT token and return access token info"""
         try:
-            # Test authentication by getting user profile
-            self.synapse_client.getUserProfile()
-            return self.synapse_client
+            import jwt
+            from jwt.exceptions import InvalidTokenError
+            from cryptography.hazmat.primitives import serialization
+
+            logger.info("Starting Synapse JWT verification")
+
+            # Decode header to get key ID
+            header = jwt.get_unverified_header(token)
+            kid = header.get('kid')
+
+            # Get JWKS and find the right key
+            jwks = await self._get_jwks()
+            key = None
+            for jwk in jwks['keys']:
+                if jwk['kid'] == kid:
+                    # Convert JWK to public key
+                    public_key = jwt.algorithms.RSAAlgorithm.from_jwk(jwk)
+                    key = public_key
+                    break
+
+            if not key:
+                logger.error(f"No key found for kid: {kid}")
+                return None
+
+            # Decode and verify token
+            decoded = jwt.decode(
+                token,
+                key,
+                algorithms=[self.algorithm],
+                audience=self.audience,
+                issuer=self.issuer
+            )
+
+            logger.info(f"JWT Payload: {decoded}")
```

Comment:
Rather than re-implementing the decode and verification flow, you could do something similar to what I've done in Schematic:

https://github.com/Sage-Bionetworks/schematic/pull/1493/files#diff-70e5179918a7449cb69edb0f7aab3388f02b24bc10a184172451e28fb71470d9